src/Uno.UI/UI/Xaml/Window/Native/NativeWindowWrapperBase.cs

```diff
@@ -91,6 +92,38 @@ public Rect VisibleBounds
 		}
 	}
 
+	/// <summary>
+	/// The same as setting <see cref="VisibleBounds"/>, <see cref="Bounds"/> and <see cref="Size"/> but makes sure the
+	/// fired events are fired only after both properties are updated "atomically"
+	/// </summary>
+	public void SetBoundsAndVisibleBounds(Rect bounds, Rect visibleBounds)
+	{
+		var applicationView = ApplicationView.GetForWindowId(Window!.AppWindow.Id);
+		var oldBounds = _bounds;
+		var oldVisibleBounds = _visibleBounds;
+		var oldVisibleBounds2 = applicationView.VisibleBounds;
+		_bounds = bounds;
+		_visibleBounds = visibleBounds;
+		applicationView.VisibleBounds = visibleBounds;
+		if (oldBounds != bounds)
+		{
+			SizeChanged?.Invoke(this, bounds.Size);
+			RaiseContentIslandStateChanged(ContentIslandStateChangedEventArgs.ActualSizeChange);
+		}
+		if (oldVisibleBounds != visibleBounds)
+		{
+			VisibleBoundsChanged?.Invoke(this, visibleBounds);
+		}
+		if (oldVisibleBounds2 != applicationView.VisibleBounds)
+		{
+			if (this.Log().IsEnabled(Uno.Foundation.Logging.LogLevel.Debug))
+			{
+				this.Log().Debug($"Updated ApplicationView visible bounds to {applicationView.VisibleBounds}");
+			}
+			applicationView.OnVisibleBoundsChanged(applicationView, null);
+		}
```

Comment:
@MartinZikmund We should probably add `Size` and `Position` to these, but I noticed that on some platforms those two are not set in the same location as Bounds and VisibleBounds so I'm not sure.